src/mlpack/methods/range_search/rs_model.hpp

```diff
@@ -55,45 +255,26 @@ class RSModel
   //! Random projection matrix.
   arma::mat q;
 
-  //! The mostly-specified type of the range search model.
-  template<template<typename TreeMetricType,
-                    typename TreeStatType,
-                    typename TreeMatType> class TreeType>
-  using RSType = RangeSearch<metric::EuclideanDistance, arma::mat, TreeType>;
-
-  // Only one of these pointers will be non-NULL.
-  //! kd-tree based range search object (NULL if not in use).
-  RSType<tree::KDTree>* kdTreeRS;
-  //! Cover tree based range search object (NULL if not in use).
-  RSType<tree::StandardCoverTree>* coverTreeRS;
-  //! R tree based range search object (NULL if not in use).
-  RSType<tree::RTree>* rTreeRS;
-  //! R* tree based range search object (NULL if not in use).
-  RSType<tree::RStarTree>* rStarTreeRS;
-  //! Ball tree based range search object (NULL if not in use).
-  RSType<tree::BallTree>* ballTreeRS;
-  //! X tree based range search object (NULL if not in use).
-  RSType<tree::XTree>* xTreeRS;
-  //! Hilbert R tree based range search object (NULL if not in use).
-  RSType<tree::HilbertRTree>* hilbertRTreeRS;
-  //! R+ tree based range search object (NULL if not in use).
-  RSType<tree::RPlusTree>* rPlusTreeRS;
-  //! R++ tree based range search object (NULL if not in use).
-  RSType<tree::RPlusPlusTree>* rPlusPlusTreeRS;
-  //! VP tree based range search object (NULL if not in use).
-  RSType<tree::VPTree>* vpTreeRS;
-  //! Random projection tree (mean) based range search object
-  //! (NULL if not in use).
-  RSType<tree::RPTree>* rpTreeRS;
-  //! Random projection tree (max) based range search object
-  //! (NULL if not in use).
-  RSType<tree::MaxRPTree>* maxRPTreeRS;
-  //! Universal B tree based range search object
-  //! (NULL if not in use).
-  RSType<tree::UBTree>* ubTreeRS;
-  //! Octree-based range search object (NULL if not in use).
-  RSType<tree::Octree>* octreeRS;
-
+ /**
+   * rSearch holds an instance of the RangeSearch class for the current
+   * treeType. It is initialized every time BuildModel is executed.
+   * We access to the contained value through the visitor classes defined above.
+   */
+  boost::variant<RSType<tree::KDTree> *,
+                 RSType<tree::StandardCoverTree> *,
+                 RSType<tree::RTree> *,
+                 RSType<tree::RStarTree> *,
+                 RSType<tree::BallTree> *,
+                 RSType<tree::XTree> *,
+                 RSType<tree::HilbertRTree> *,
+                 RSType<tree::RPlusTree> *,
+                 RSType<tree::RPlusPlusTree> *,
+                 RSType<tree::VPTree> *,
+                 RSType<tree::RPTree> *,
+                 RSType<tree::MaxRPTree> *,
+                 RSType<tree::UBTree> *,
+                 RSType<tree::Octree> *> rSearch;
```

Comment:
This is a minor style comment---no space is needed between the type and the pointer, so `RSType<tree::UBTree>*` not `RSType<tree::UBTree> *` please.